prime-router/docs/proposals/0004-git-flow-and-versioning.md

```diff
@@ -0,0 +1,63 @@
+
+> `feature_branch(es)` --> `master` --> `production`
+
+While this flow has served its purpose for the testing and validation of software changes before deployment to production, a simpler workflow that allows for the versioning of the platform is a necessary next step in its maturation.
+
+# Goals
+
+The goal of this proposal is to 1) simplify the flow of code from development to production and 2) to employ versioning for a clear lineage and changelog throughout the platform's lifecycle.
+
+# Proposal
+
+## Versioning
+
+Versioning of the PRIME Data Hub platform can be discerned in two ways - this proposal will focus on the first point:
+  - The overall version of the platform
+  - The version of the platform's constituent components
+
+The overall platform should be versioned in a temporally distinct way, yet allowing for multiple production releases within a single calendar day.  The proposed format is as follows:
+
+> YYYY.0M.0D_MICRO
+
+|YYYY|0M|0D|MICRO
+|---|---|---|---|
+|Full year|Zero-padded month|Zero-padded day|Release number|
+|2019, 2020, ...|01, 10, ...|01, 30, ...|0, 1, 2, ...
+
+The proposed versioning scheme is based off of `Calendar Versioning`<sup>1</sup>.  A separate proposal may leverage `Semantic Versioning`<sup>2</sup> to add granularity to the individual components within the PRIME Data Hub platform.
+
+## Git Flow
+
+To simplify the flow of code from development to the test and production environments, the proposed flow will be:
+
+> `feature_branch(es)` --> `master`
+
+The `master` branch will remain protected in that:
+  - It cannot be pushed to directly
+  - A pull request must be submitted from the feature branch
+  - The pull request must be approved by ≥ 1 reviewer prior to being merged into the `master` branch
+
+Upon the creation of a pull request, the `test_build` workflow will be triggered to build and test the changes that would result from the pull request being merged.
+
+Upon merge, the GitHub actions `test_release` workflow will be triggered to deploy to the `prime-data-hub-test` resource group within Azure.  The merged changes should then be thoroughly tested within the cloud environment.
+
+This proposed flow is based off of the `GitHub Flow`<sup>3</sup>.
+
+## Production Release
+
+When a collection of features is deemed ready for production deployment, a release should be published.  Upon publish, the `prod_release` workflow will be triggered to deploy to the `prime-data-hub-prod` resource group within Azure.
+
+To create a release, navigate to the `Create a new release` link on the right-hand side of the main GitHub repository page.  Enter the appropriate version, title, and description and click `Publish release`.  The title should be a succinct overview of the theme(s) surrounding each release.  The description should contain a detailed change log describing the changes that were made to the platform.
```

Comment:
How are failures handled in either case? If a deployment fails, to re-run another one? Will we have a manual GitHub Action trigger to re-run a deployment?

What about hotfixes to production in the event something was merged to test and a PR was merged to test? Should we be treating as everything merged into master is safe to deploy to production, and there will never be a state when master is unsafe to deploy?

Can anyone run a deployment? Only a developer who has changes in master? Select team members?

How often should we be releasing code to production?

Reply:
So should the release comment be a copy of the release notes. Or point to the release notes section?

Reply:
Is there a use for the pre-release feature?

Reply:
> How are failures handled in either case? If a deployment fails, to re-run another one? Will we have a manual GitHub Action trigger to re-run a deployment?

If a deployment fails because of a bad build, this would be caught going into the test environment.  Code updates can be pushed to then trigger the build to run again.  In the case of a production deployment failing, we have the ability to go in and re-run the exact version of the workflow that ran via GitHub actions.

> 
> What about hotfixes to production in the event something was merged to test and a PR was merged to test? Should we be treating as everything merged into master is safe to deploy to production, and there will never be a state when master is unsafe to deploy?

With the assumptions that changes are tested first locally and then in the testing environment - code in the master branch should be safe to deploy to production.  If a developer is testing in the test environment and finds a change to not produce the desired effect, then it should be communicated to the team to not publish any releases until that person has either reverted the change or pushed and tested a fix.

> 
> Can anyone run a deployment? Only a developer who has changes in master? Select team members?

I'm open to suggestions on this.  Currently, anybody in the CDCgov organization with write access can publish a release.  Do we want that process to be different?

> 
> How often should we be releasing code to production?

I'm also open to suggestions here as well.  We currently organically find a point-in-time where a set of features are ready and do an *ad hoc* deployment

Reply:
> So should the release comment be a copy of the release notes. Or point to the release notes section?

My thought was that the release comment would serve as the container for release notes, but if you prefer a separate document for release notes it can serve as a pointer instead.

Reply:
> Is there a use for the pre-release feature?

If we operate under the assumption that code making it to `master` is production-ready - then I do not think we will have a need tor a pre-release.  Alternatively, each release could start as a pre-release so that a second person can review it prior to deploying to production.  Thoughts?

Reply:
> > Can anyone run a deployment? Only a developer who has changes in master? Select team members?
> 
> I'm open to suggestions on this. Currently, anybody in the CDCgov organization with write access can publish a release. Do we want that process to be different?
> 

This is a good discussion. I think that publishing a release is an equivalent power to being able to commit to master.

Reply:
> If a deployment fails because of a bad build, this would be caught going into the test environment. Code updates can be pushed to then trigger the build to run again. In the case of a production deployment failing, we have the ability to go in and re-run the exact version of the workflow that ran via GitHub actions.

I would disagree with this statement to some effect. While deployment errors should be caught going to a test environment, that's not always the case. Sometimes, especially in the state of the project we're in now, it's possible there are slight differences between environments that cause issues, so a deployment to test might succeed but a deployment to prod fails.

Regardless, I think ensuring we have the ability to re-run a deployment in the event of a failure is essential, and we should get that out of the box with GitHub Actions. I would also suggest we add a restriction to our production deployment that the test deployment's last run must have succeeded for the production deployment to run.

Reply:
> With the assumptions that changes are tested first locally and then in the testing environment - code in the master branch should be safe to deploy to production. If a developer is testing in the test environment and finds a change to not produce the desired effect, then it should be communicated to the team to not publish any releases until that person has either reverted the change or pushed and tested a fix.

Agree with this, and I think this can be mitigated if we have a more limited set of team members who can run deploys. I would agree with @RickHawesUSDS that only team members with direct commit access to master. This would be either team members who have "Maintain" or "Admin" access to the repository.